File: datapull.rb
#!/usr/bin/env ruby
require 'mongo'
require 'rest_client'
require 'json'

# Setup Mongo
conn = Mongo::Connection.new
db = conn['hub']
collection = db['repos']

# Clear out old collection
collection.remove

GH_API = 'https://api.github.com/'
SEARCH = 'legacy/repos/search/'

SEARCH_RESOURCE = GH_API + SEARCH + 'mongodb'

page = 0

loop do
  puts "Processing page #{page}"
  url = SEARCH_RESOURCE + "?start_page=#{page}" 
  result = RestClient.get url, {:accept => :json}
  rh = JSON.load(result)
  if rh['repositories'].length > 0
    collection.insert(rh['repositories'])
    page += 1
  else
    break
  end
end

File: datapull.py
# Note: this expects python 3

import json
import urllib.request

from datetime import datetime
from pprint import pprint

import pymongo
from pymongo.errors import (DuplicateKeyError,
                            OperationFailure)

c = pymongo.Connection()
db = c.githubdata
coll = db.repos

# github's paging seems to return dupes...
coll.ensure_index([('username', pymongo.ASCENDING),
                   ('name', pymongo.ASCENDING)], unique=True)

fmt = "%Y-%m-%dT%H:%M:%S"

page = 0

while True:

    print(page)
    response = urllib.request.urlopen('https://api.github.com/legacy/'
                                      'repos/search/:mongodb?start_page'
                                      '=%d' % (page,))

    data = response.read()
    js = json.loads(data.decode())
    # github returns an empty array for 'repositories'
    # when there are no more results.
    if js['repositories']:
        for repo in js['repositories']:
            # Convert date strings to real timestamps.
            for field in ('created', 'created_at', 'pushed', 'pushed_at'):
                # TODO: This sucks ass but skip timezone offsets for the time being.
                repo[field] = datetime.strptime(repo[field].rsplit('-', 1)[0], fmt)
            try:
                # Update so we get updated followers and watchers count.
                res = coll.update({'username': repo['username'],
                                   'name': repo['name']},
                                  repo, upsert=True, safe=True)
                if not res['updatedExisting']:
                    pprint(repo)
            except (DuplicateKeyError, OperationFailure):
                pass

        page += 1

    else:
        break

